dalm/datasets/reading_comprehension_generation/synthetic_based.py

```diff
@@ -0,0 +1,104 @@
+    return [
+        {
+            "role": "system",
+            "content": "You are a helpful and meticulous instruction following question and answer making chatbot. Please refrain from acknowledgments, additions or niceties of any sort",
+        },
+        {"role": "user", "content": prompt},
+    ]
+
+
+def generate_synthetic_data(model_pipeline, text, generation_params):
+    prompt = gen_prompt(text)
+    prompt = model_pipeline.tokenizer.apply_chat_template(prompt, tokenize=False, add_generation_prompt=True)
+    outputs = model_pipeline(prompt, **generation_params)
+
+    return outputs[0]["generated_text"]
+
+
+def generate_synthetic_dataset(
+    model_name,
+    input_directory,
+    state_file=None,
+    chunk=False,
+    context_length=2048,
+    generation_params={
+        "max_new_tokens": 600,
+        "do_sample": True,
+        "temperature": 0.7,
+        "top_k": 5,
+        "top_p": 0.95,
+        "return_full_text": False,
+    },
+):
+    model_pipeline = pipeline("text-generation", model=model_name, torch_dtype=torch.bfloat16, device_map="auto")
+
+    input_files = list_dir(input_directory)
+
+    if chunk:
+        tokenizer = model_pipeline.tokenizer
+        tokens = tokenizer.apply_chat_template(gen_prompt(""), tokenize=False, add_generation_prompt=True)
+        CONSTANT = len(tokenizer(tokens)["input_ids"])
+        k = context_length - CONSTANT
+
+    if state_file:
+        if os.path.exists(state_file):
+            with open(state_file, "rb") as f:
+                state = pickle.load(f)
+        else:
+            state = {"processed_files": []}
+            pickle.dump(state, open(state_file, "wb"))
+
+    for file, text in input_files:
+        if state and file in state["processed_files"]:
+            continue
+
+        if chunk:
+            for chunk_ in text_chunker(text, tokenizer, k):
+                gen_text = generate_synthetic_data(model_pipeline, chunk_, generation_params)
+                yield gen_text, chunk_
+        else:
+            gen_text = generate_synthetic_data(model_pipeline, text, generation_params)
+            yield gen_text, text
+
+        if state:
+            state["processed_files"].append(file)
+            pickle.dump(state, open(state_file, "wb"))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", type=str, required=True)
+    parser.add_argument("--input_directory", type=str, required=True)
+    parser.add_argument("--output_directory", type=str, required=True)
+    parser.add_argument("--state_file", type=str, required=False, default="rc_generation_state.pkl")
+    parser.add_argument("--context_length", type=int, default=2048)
+    parser.add_argument("--chunk", action="store_true")
+
+    args = parser.parse_args()
```

Comment:
To fail-fast, it would be good to check if all required args were passed:

```
if not args.model_name:
    parser.error("--model_name is a required argument")
... etc
```

Otherwise what might happen is that it takes several minutes to run, then throws an exception because the `output_directory` argument was missing.  This would be frustrating for the user.

Reply:
Nevermind, `argparse` should take care of this for us as long as we mark all args as `required=True` (see other comment regarding `model_name`)